OpenSim/Common/TableSource.h

```diff
@@ -0,0 +1,214 @@
+    const Table& getTable() const {
+        return _table;
+    }
+
+    /** Replace the existing TimeSeriesTable_ this TableSource_ currently 
+    holds.                                                                    
+
+    \throws KeyNotFound If table provided does not have column-labels.        */
+    void setTable(const Table& table) {
+        _table = table;
+        auto& columnOutput = updOutput("column");
+        columnOutput.clearChannels();
+        for(const auto& columnLabel : _table.getColumnLabels())
+            columnOutput.addChannel(columnLabel);
+    }
+
+    /// @}
+
+protected:
+    /** Retrieve value of a column at a given time(implicit in the State 
+    provided). Linear interpolation is performed if the TimeSeriesTable_ does
+    not contain an entry for the time mentioned by the state.
+
+    \throws EmptyTable If the TimeSeriesTable_ this TableSource_ holds is 
+                       currently empty.
+    \throws TimeOutOfRange If the time specified by the State is either less 
+                           than the smallest timestamp or greater than the 
+                           largest timestamp in the TimeSeriesTable_.
+    \throws KeyNotFound If TimeSeriesTable_ does not have column-labels.      */
+    ET getColumnAtTime(const SimTK::State& state, 
+                       const std::string& columnLabel) const {
+        OPENSIM_THROW_IF(_table.getNumRows() == 0, EmptyTable);
+        const auto& timeCol = _table.getIndependentColumn();
+        const auto time = state.getTime();
+        OPENSIM_THROW_IF(time < timeCol.front() ||
+                         time > timeCol.back(),
+                         TimeOutOfRange, 
+                         time, timeCol.front(), timeCol.back());
+
+        const auto colInd = _table.getColumnIndex(columnLabel);
+        auto lb = std::lower_bound(timeCol.begin(), timeCol.end(), time);
+        if(lb == timeCol.begin())
+            return _table.getMatrix().getElt(0, colInd);
+        else if(lb == timeCol.end())
+            return _table.getMatrix().getElt(timeCol.size() - 1, colInd);
+        else if(*lb == time)
+            return _table.getMatrix().getElt(lb - timeCol.begin(), colInd);
+        else {
+            auto prevTime = *(lb - 1);
+            auto nextTime = *lb;
+            auto prevElt = _table.getMatrix().getElt(lb - 1 - timeCol.begin(), 
+                                                     colInd);
+            auto nextElt = _table.getMatrix().getElt(lb - timeCol.begin(), 
+                                                     colInd);
+            auto elt = ((time - prevTime) / (nextTime - prevTime)) * 
+                       (nextElt - prevElt) + prevElt;
+            return elt;
+        }
+    }
+
+    /** Retrieve a row of the TimeSeriesTable_ at a given time (specified by the
+    state). Linear interpolation is performed if the TimeSeriesTable_ does not
+    have an entry for the time mentioned by the state.
```

Comment:
Is time included in the Vector? Perhaps include a `\returns` comment explicitly stating that a Vector is returned whose length is equal to… :snowboarder: